cou/steps/plan.py

```diff
@@ -375,6 +379,55 @@ async def generate_plan(analysis_result: Analysis, args: CLIargs) -> UpgradePlan
     return plan
 
 
+async def filter_hypervisors_machines(args: CLIargs, analysis_result: Analysis) -> set[Machine]:
+    """Filter the hypervisors to generate plan and upgrade.
+
+    :param args: CLI arguments
+    :type args: CLIargs
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :return: hypervisors filtered to generate plan and upgrade.
+    :rtype: set[Machine]
+    """
+    hypervisors_machines = await _get_upgradable_hypervisors_machines(args.force, analysis_result)
+
+    if cli_machines := args.machines:
+        return {machine for machine in hypervisors_machines if machine.machine_id in cli_machines}
+
+    if cli_hostnames := args.hostnames:
+        return {machine for machine in hypervisors_machines if machine.hostname in cli_hostnames}
+
+    if cli_azs := args.availability_zones:
+        return {machine for machine in hypervisors_machines if machine.az in cli_azs}
+
+    return hypervisors_machines
+
+
+async def _get_upgradable_hypervisors_machines(
+    cli_force: bool, analysis_result: Analysis
+) -> set[Machine]:
+    """Get the hypervisors that are possible to upgrade.
+
+    :param cli_force: If force is used, it gets all hypervisors, otherwise just the empty ones
+    :type cli_force: bool
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :return: Set of nova-compute units to upgrade
+    :rtype: set[Machine]
+    """
+    nova_compute_units = {
```

Comment:
Why we are creating set and converting it to list in 428, I think we should do list here instead. The set does not bring any benefit, since units are unique.

Reply:
It was a way to be sure that we wouldn't have duplicate machines, but I think you are right. I'll change the data structure to be lists